modules/apps/frontend-icons/frontend-icons-admin-web/src/main/java/com/liferay/frontend/icons/admin/web/internal/portal/settings/configuration/admin/display/IconsAdminPortalSettingsConfigurationScreenContributor.java

```diff
@@ -0,0 +1,227 @@
+	}
+
+	@Override
+	public ServletContext getServletContext() {
+		return _servletContext;
+	}
+
+	@Override
+	public void setAttributes(
+		HttpServletRequest httpServletRequest,
+		HttpServletResponse httpServletResponse) {
+
+		IconsAdminConfiguration IconsAdminConfiguration = null;
+
+		try {
+			IconsAdminConfiguration =
+				_configurationProvider.getCompanyConfiguration(
+					IconsAdminConfiguration.class,
+					CompanyThreadLocal.getCompanyId());
+		}
+		catch (PortalException portalException) {
+			ReflectionUtil.throwException(portalException);
+		}
+
+		httpServletRequest.setAttribute(
+			IconsAdminConfiguration.class.getName(), IconsAdminConfiguration);
+
+
+		Set<Map.Entry<String, List<IconResource>>> entries =
+			_iconResourcesMap.entrySet();
+
+
+		List<String> ids = new ArrayList<>();
+
+		for (Map.Entry<String, List<IconResource>> entry : entries) {
+			List<IconResource> iconResources = entry.getValue();
+
+			IconResource iconResource = iconResources.get(0);
+
+			ids.add(iconResource.getId());
+		}
+
+		httpServletRequest.setAttribute("iconIds", ids);
+	}
+
+	@Reference
+	private ConfigurationProvider _configurationProvider;
+
+	@Reference(
+		target = "(osgi.web.symbolicname=com.liferay.frontend.icons.admin.web)",
+		unbind = "-"
+	)
+	private ServletContext _servletContext;
+
+
+	@Reference(
+		cardinality = ReferenceCardinality.MULTIPLE,
+		policy = ReferencePolicy.DYNAMIC
+	)
+	protected void addIconResourcesContributor(
+		IconResourcesContributor iconResourcesContributor) {
+
+		Lock lock = _readWriteLock.writeLock();
+
+		lock.lock();
+
+		try {
+			for (IconResource iconResource :
+					iconResourcesContributor.getIconResources()) {
+
+				_addIconResource(iconResource);
+			}
+		}
+		finally {
+			lock.unlock();
+		}
+	}
+
+	protected void removeIconResourcesContributor(
+		IconResourcesContributor iconResourcesContributor) {
+
+		Lock lock = _readWriteLock.writeLock();
+
+		try {
+			lock.lock();
+
+			for (IconResource iconResource :
+					iconResourcesContributor.getIconResources()) {
+
+				_removeIconResource(iconResource);
+			}
+		}
+		finally {
+			lock.unlock();
+		}
+	}
+
+	private void _addIconResource(IconResource iconResource) {
+		String id = iconResource.getId();
+
+		List<IconResource> iconResources = _iconResourcesMap.get(id);
+
+		if (iconResources == null) {
+			iconResources = new ArrayList<>();
+
+			_iconResourcesMap.put(id, iconResources);
+		}
+
+		int i;
+
+		for (i = 0; i < iconResources.size(); i++) {
+			IconResource existingIconResource = iconResources.get(i);
+
+			if (existingIconResource.getPriority() >
+					iconResource.getPriority()) {
+
+				break;
+			}
+		}
+
+		iconResources.add(i, iconResource);
+	}
+
+	private void _removeIconResource(IconResource iconResource) {
+		List<IconResource> iconResources = _iconResourcesMap.get(
+			iconResource.getId());
+
+		if (iconResources == null) {
+			return;
+		}
+
+		iconResources.remove(iconResource);
+	}
```

Comment:
This is all duplicated in the DynamicInclude as well. Is there an easy way to share this code between the two?

Reply:
You can create a `XxxUtil` class with static methods, for example. Or a `XxxHelper` that can be instantiated (if static methods are not valid and you need to maintain some state between calls).